src/substrate/filters/enricher.ts

```diff
@@ -101,55 +110,103 @@ export async function Enrich(
+          ? await api.query.staking.erasStakers.keysAt(hash,currentEra)
+          : await api.query.staking.stakers.keysAt(hash, currentEra) ; 
+
+        const nextElected = keys.length ? keys.map((key) => key.args[1] as AccountId): validators['nextElected'];
+        
+        // get current stashes
+        const stashes = await api.query.staking.validators.keysAt(hash);
+
+        // find the waiting validator
+        const nextElectedStr = nextElected.map((v) => v.toString())
+        // const stashesStr =stashes.map((v) => v.args[0].toString())
+        const stashesStr =stashes.filter((v) => v.args.length).map((v) => v.args[0].toString())
+        
+        const waiting = stashesStr.filter((v) => !nextElectedStr.includes(v));
+
+
+        // get validators current era reward points
+        const validatorEraPoints: EraRewardPoints = await currentPoints(api, currentEra, hash, validators) as EraRewardPoints;
+        const eraPointsIndividual = validatorEraPoints.individual.toJSON();
+
         const validatorInfo = {};
-        electedInfo.info.forEach(async ({ accountId, controllerId, validatorPrefs, rewardDestination }) => {
-          const commissionPer = (Number)(validatorPrefs.commission || new BN(0)) / 10_000_000;
-          const key = accountId.toString();
-          validatorInfo[key] = {}
+
+        for(let validator of validators){
+          const key = validator.toString();
+
+          // eraValidatorPrefs does not return comission prior to 3139800
+          const preference = api.query.staking.erasValidatorPrefs
+            ? await api.query.staking.erasValidatorPrefs.at(hash, currentEra,key)
+            : await api.query.staking.validators.at(hash, key);
+
+          const commissionPer =  (Number)(preference.commission || new BN(0)) / 10_000_000;
+
+          const rewardDestination = await api.query.staking.payee.at(hash,key);
+          const controllerId = await api.query.staking.bonded.at(hash, key);
+          let nextSessionKeysOpt;
+
+          try{
+            // to get keys of old blocks
+            nextSessionKeysOpt = await api.query.session.nextKeys(hash,key);
+            // there is queuedKeys function as well not sure which one to use
+            // nextSessionKeysOpt = await api.query.session.queuedKeys(hash,key);
+          }
+          catch(e){
+            // for new blocks after 3139200
+            nextSessionKeysOpt = await api.query.session.nextKeys.at(hash,key);
+          }
+          
+          const nextSessionKeys = nextSessionKeysOpt.isSome
+            ? nextSessionKeysOpt.unwrap()
+            : []
+
           validatorInfo[key] = {
             commissionPer,
-            controllerId,
-            rewardDestination,
-            nextSessionIds: await (await api.query.session.nextKeys(key)).toString(),
-            eraPoints: Number(eraPointsIndividual[key])
+            controllerId: controllerId.toString() == ''? key: controllerId.toString(),
```

Comment:
`controllerId` is an option, so why are we converting to string to check its value?